jax/_src/nn/functions.py

```diff
@@ -765,3 +767,172 @@ def hard_silu(x: ArrayLike) -> Array:
   return x_arr * hard_sigmoid(x_arr)
 
 hard_swish = hard_silu
+
+def _get_large_negative(dtype):
+  dtype_max = jnp.finfo(dtype).max
+  return jnp.asarray(-0.7 * dtype_max, dtype=dtype)
+
+def _get_causal_mask(T, dtype):
+  large_negative_number = _get_large_negative(dtype)
+  col_idx = jnp.tile(jnp.arange(T)[jnp.newaxis, :], [T, 1])
+  row_idx = jnp.tile(jnp.arange(T)[:, jnp.newaxis], [1, T])
+  mask = (row_idx < col_idx).astype(dtype) * large_negative_number
+  mask = mask[jnp.newaxis, jnp.newaxis, :, :]
+  return mask
+
+def _dot_product_attention_xla(
+    query: ArrayLike,
+    key: ArrayLike,
+    value: ArrayLike,
+    bias: ArrayLike | None,
+    mask: ArrayLike | None,
+    is_causal: bool,
+    scale: float):
+  logits_dtype = jnp.promote_types(query.dtype, jnp.float32)
+  # Compute the attention logits
+  logits = jnp.einsum('BTNH,BSNH->BNTS', query, key,
+                      preferred_element_type=logits_dtype)
+
+  # Logits scaling
+  scale = jnp.array(scale, dtype=logits.dtype)
+  logits = jnp.multiply(logits, scale)
+
+  # Bias
+  if bias is not None:
+    logits = logits + bias.astype(logits.dtype)
+
+  # Mask
+  if mask is not None:
+    assert mask.dtype == jnp.bool_
+    large_negative_number = _get_large_negative(logits.dtype)
+    padded_logits = jnp.where(mask, logits, large_negative_number)
+  else:
+    padded_logits = logits
+
+  if is_causal:
+    T, S = query.shape[-3], key.shape[-3]
+    assert T == S, f'T should equal to S, but got {T} vs {S}'
```

Comment:
This assertion isn't necessary

Reply:
Do you mean using causal mask doesn't require T and S to be same?

Reply:
Or somewhere else has already ensured this?

Reply:
Ping @chr1sj0nes 👋

Reply:
I think it is possible to perform causal attention with T != S.

Reply:
Thx for clarification. Removed the unnecessary checks.